nautobot/users/tests/test_password_ui.py

```diff
@@ -0,0 +1,89 @@
+from django.contrib.sessions.middleware import SessionMiddleware
+from django.test import override_settings, RequestFactory, TestCase
+from django.urls import reverse
+
+from social_django.utils import load_strategy, load_backend
+
+User = get_user_model()
+
+
+class PasswordUITest(TestCase):
+    def tearDown(self):
+        self.client.logout()
+
+    def test_change_password_enabled(self):
+        """
+        Check that a Django-authentication-based user is allowed to change their password
+        """
+        non_sso_user = User.objects.create_user(username="non_sso_user", is_superuser=True)
+        self.client.force_login(non_sso_user)
+        profile_response = self.client.get(reverse("user:profile"))
+        preferences_response = self.client.get(reverse("user:preferences"))
+        api_tokens_response = self.client.get(reverse("user:token_list"))
+        for response in [profile_response, preferences_response, api_tokens_response]:
+            self.assertIn("Change Password", str(response.content))
+
+        # Check GET change_password functionality
+        get_response = self.client.get(reverse("user:change_password"))
+        self.assertIn("New password confirmation", str(get_response.content))
+
+        # Check POST change_password functionality
+        post_response = self.client.post(
+            reverse("user:change_password"),
+            data={
+                "old_password": "foo",
+                "new_password1": "bar",
+                "new_password2": "baz",
+            },
+        )
+        self.assertIn("The two password fields", str(post_response.content))
+
+    @override_settings(
+        AUTHENTICATION_BACKENDS=[
+            "social_core.backends.google.GoogleOAuth2",
+            "nautobot.core.authentication.ObjectPermissionBackend",
+        ]
+        # unnecessary
+    )
+    def test_change_password_disabled(self):
+        """
+        Mock an SSO-authenticated user, log them in by force and check that the change
+        password functionality isn't visible in the UI or available server-side
+        """
+        sso_user = User.objects.create_user(username="sso_user", is_superuser=True)
+        self.request_factory = RequestFactory()
+        self.request = self.request_factory.get("/")
+        SessionMiddleware(lambda: None).process_request(self.request)
+        strategy = load_strategy(request=self.request)
+        backend = load_backend(strategy=strategy, name="google-oauth2", redirect_uri="/")
+
+        # Mock an authenticated SSO pipeline
+        with mock.patch("social_core.backends.base.BaseAuth.pipeline", return_value=sso_user):
+            result = strategy.authenticate(backend=backend, response=mock.Mock())
+            self.assertEqual(result, sso_user)
+            self.assertEqual(result.backend, "social_core.backends.google.GoogleOAuth2")
+            self.assertTrue(sso_user.is_authenticated)
+            self.client.force_login(sso_user)
```

Comment:
The `strategy.authenticate` wasn't actually logging in the user. That was why the test wasn't working as expected yesterday. Adding `force_login` made the client session testable. I'll investigate why `strategy.authenticate` didn't actually log the user in. Think it might be the mocking somewhere. I'll check it out.

Reply:
OK so it looks like `authenticate()` doesn't actually persist the user in the request. `login()` does:

Django source login() docstring:

`
def login(request, user, backend=None):
    """
    Persist a user id and a backend in the request. This way a user doesn't
    have to reauthenticate on every request. Note that data set during
    the anonymous session is retained when the user logs in.
    """
`